components/eamxx/src/physics/iop_forcing/eamxx_iop_forcing_process_interface.cpp

```diff
@@ -0,0 +1,546 @@
+    auto s_ref_p_int = ekat::scalarize(ref_p_int);
+    Kokkos::parallel_for(Kokkos::TeamVectorRange(team, num_levs+1), [&](const int& k) {
+      s_ref_p_int(k) = hyai(k)*ps0 + hybi(k)*ps_i;
+      if (k < num_levs) {
+        s_ref_p_mid(k) = hyam(k)*ps0 + hybm(k)*ps_i;
+      }
+    });
+    team.team_barrier();
+    ColOps::compute_midpoint_delta(team, num_levs, ref_p_int, ref_p_del);
+    team.team_barrier();
+
+    if (iop_dosubsidence) {
+    // Compute subsidence due to large-scale forcing
+      advance_iop_subsidence(team, num_levs, dt, ps_i, ref_p_mid, ref_p_int, ref_p_del, omega, ws, u_i, v_i, T_mid_i, Q_i);
+    }
+
+    // Update T and qv according to large scale forcing as specified in IOP file.
+    advance_iop_forcing(team, num_levs, dt, divT, divq, T_mid_i, qv_i);
+
+    if (iop_coriolis) {
+      // Apply coriolis forcing to u and v winds
+      iop_apply_coriolis(team, num_levs, dt, target_lat, u_ls, v_ls, u_i, v_i);
+    }
+
+    // Release WS views
+    ws.release_many_contiguous<3>({&ref_p_mid, &ref_p_int, &ref_p_del});
+  });
+  Kokkos::fence();
+
+  // Nudge the domain based on the domain mean
+  // and observed quantities of T, Q, u, and v
+  if (iop_nudge_tq or iop_nudge_uv) {
+    // Compute domain mean of qv, T_mid, u, and v
+    const auto qv_mean = m_buffer.qv_mean;
+    const auto t_mean  = m_buffer.t_mean;
+    const auto u_mean  = m_buffer.u_mean;
+    const auto v_mean  = m_buffer.v_mean;
+
+    const auto qv_mean_h = Kokkos::create_mirror_view(qv_mean);
+    const auto t_mean_h  = Kokkos::create_mirror_view(t_mean);
+    const auto u_mean_h  = Kokkos::create_mirror_view(u_mean);
+    const auto v_mean_h  = Kokkos::create_mirror_view(v_mean);
+
+    const auto num_global_cols = m_grid->get_num_global_dofs();
+    for (int k=0; k<m_num_levs; ++k) {
+      if (iop_nudge_tq){
+        Real& qv_mean_k = qv_mean_h(k/Pack::n)[k%Pack::n];
+        Real& t_mean_k = t_mean_h(k/Pack::n)[k%Pack::n];
+        Kokkos::parallel_reduce("compute_domain_means_tq",
+                                m_num_cols,
+                                KOKKOS_LAMBDA (const int icol, Real& q_sum, Real& t_sum) {
+          q_sum += qv(icol, k/Pack::n)[k%Pack::n];
+          t_sum += T_mid(icol, k/Pack::n)[k%Pack::n];
+        },  qv_mean_k, t_mean_k);
+
+        m_comm.all_reduce(&qv_mean_k, 1, MPI_SUM);
+        m_comm.all_reduce(&t_mean_k, 1, MPI_SUM);
+
+        qv_mean_k /= num_global_cols;
+        t_mean_k /= num_global_cols;
+      }
+      if (iop_nudge_uv){
+        Real& u_mean_k = u_mean_h(k/Pack::n)[k%Pack::n];
+        Real& v_mean_k = v_mean_h(k/Pack::n)[k%Pack::n];
+        Kokkos::parallel_reduce("compute_domain_means_uv",
+                                m_num_cols,
+                                KOKKOS_LAMBDA (const int icol, Real& u_sum, Real& v_sum) {
+          u_sum += horiz_winds(icol, 0, k/Pack::n)[k%Pack::n];
+          v_sum += horiz_winds(icol, 1, k/Pack::n)[k%Pack::n];
+        }, u_mean_k, v_mean_k);
```

Comment:
I think you may need a fence. here. I'm not 100% sure though, the ||reduce may already put one in for you. But we may as well be safe and make sure you don't start the MPI ops without finishing the kernel.

Reply:
A ||reduce with a scalar reduction type implies a fence (a rank 0 view reduction type does not imply a fence), but I'm not sure if that is guaranteed to always be the behavior. Would definitely be safer to fence.